srai/joiners/intersection_joiner.py

```diff
@@ -0,0 +1,58 @@
+"""
+Intersection Joiner.
+
+This module contains intersection joiner implementation.
+
+"""
+
+import geopandas as gpd
+import pandas as pd
+
+from srai.joiners import JoinerResult
+
+
+class IntersectionJoiner:
+    """
+    Intersection Joiner.
+
+    Intersection Joiner allows to join two GeoDataFrames and find all
+    overlapping geometries. It does not apply any grouping or
+    aggregation.
+
+    """
+
+    def join(self, regions: gpd.GeoDataFrame, features: gpd.GeoDataFrame) -> JoinerResult:
+        """
+        Join features to regions.
+
+        Args:
+            regions (gpd.GeoDataFrame): regions with which features are joined
+            features (gpd.GeoDataFrame): features to be joined
+
+        Returns:
+            JoinerResult object, which contains 3 GeoDataFrames:
+            * original regions
+            * original features
+            * intersection of regions and features, which contains a MultiIndex and a geometry with
+              the intersection
+
+        """
+        joined_parts = []
+
+        for _, single in features.groupby(features["geometry"].geom_type):
+            joined_parts.append(
+                gpd.overlay(
+                    single[["geometry"]].reset_index().rename(columns={"index": "features_id"}),
+                    regions[["geometry"]].reset_index().rename(columns={"index": "region_id"}),
+                    how="intersection",
+                    keep_geom_type=False,
+                ).set_index(["region_id", "features_id"])
+            )
+
+        joined = gpd.GeoDataFrame(pd.concat(joined_parts, ignore_index=False))
+
+        return JoinerResult(
```

Comment:
Why do we want to return all 3 gdfs here?

Reply:
If we process the whole pipeline probably it is not needed. But we decided to return all features here, and I followed @RaczeQ's suggestion to return it in this form (https://github.com/srai-lab/srai/issues/55#issuecomment-1296146540)